### ISSUE TITLE
clusterctl upgrade `Timed out waiting for all Machines to exist`

### ISSUE THREAD

**Issue by cahillsf** [kubernetes-sigs/cluster-api#11209]
### Which jobs are flaking?

`periodic-cluster-api-e2e-main`

### Which tests are flaking?

`When testing clusterctl upgrades (v0.3=>v1.5=>current) Should create a management cluster and then upgrade all the providers`

specifically looking at this pattern:
* [36 failures](https://storage.googleapis.com/k8s-triage/index.html?date=2024-09-02&text=Timed%20out%20waiting%20for%20all%20Machines%20to%20exist&job=.*periodic-cluster-api-e2e.*&test=.*clusterctl%20upgrades): `Timed out waiting for all Machines to exist`
  
  * Component: unknown
  * Branches:
    * main
    * release-1.8
    * release-1.7

### Since when has it been flaking?

for quite some time

### Testgrid link

https://testgrid.k8s.io/sig-cluster-lifecycle-cluster-api#capi-e2e-main

### Reason for failure (if possible)

TLDR: seems like there is an issue in the Docker controller in creating the worker machines with many logs like:
`dockermachine_controller.go:220] controllers/DockerMachine/DockerMachine-controller "msg"="failed to create worker DockerMachine: timed out waiting for the condition, cleaning up so we can re-provision from a clean state" `
The failures follow different patterns

I think increasing `clusterctl-upgrade/wait-worker-nodes` e2e interval is a good first step as some of the examples below show the retry feedback in creating the `DockerMachine`s appearing to resolve but the tests time out

the errors don't provide much info, but since the issue seems to stem from the docker container runtime `RunContainer`(https://github.com/kubernetes-sigs/cluster-api/blob/879617dcc25735ef734d33adad9618707d43a95b/test/infrastructure/docker/internal/docker/manager.go#L182) call, we could explore passing in an output Writer here as the[ third parameter](https://github.com/kubernetes-sigs/cluster-api/blob/879617dcc25735ef734d33adad9618707d43a95b/test/infrastructure/container/docker.go#L375) to get output directly from the container?

## Example failure 1 (expected 2, found 1) 
(`When testing clusterctl upgrades (v0.3=>v1.5=>current) Should create a management cluster and then upgrade all the providers`) [prow link](https://prow.k8s.io/view/gs/kubernetes-jenkins/logs/periodic-cluster-api-e2e-main/1829718349903302656)

[capd controller logs](https://storage.googleapis.com/kubernetes-jenkins/logs/periodic-cluster-api-e2e-main/1829718349903302656/artifacts/clusters/clusterctl-upgrade-management-52n6gm/logs/capd-system/capd-controller-manager/capd-controller-manager-fb4b578f9-7pccs/manager.log) show: 

`I0831 03:46:04.834929       1 dockermachine_controller.go:220] controllers/DockerMachine/DockerMachine-controller "msg"="failed to create worker DockerMachine: timed out waiting for the condition, cleaning up so we can re-provision from a clean state" "cluster"="clusterctl-upgrade-workload-q53g9g" "docker-cluster"="clusterctl-upgrade-workload-q53g9g" "docker-machine"={"Namespace":"clusterctl-upgrade","Name":"clusterctl-upgrade-workload-q53g9g-md-0-lfnm6"} "machine"="clusterctl-upgrade-workload-q53g9g-md-0-5fb876f4bc-cq8k6"`

in this case the CP machine is fully initialized, but the [MD worker machine status](https://storage.googleapis.com/kubernetes-jenkins/logs/periodic-cluster-api-e2e-main/1829718349903302656/artifacts/clusters/clusterctl-upgrade-management-52n6gm/resources/clusterctl-upgrade/DockerMachine/clusterctl-upgrade-workload-q53g9g-md-0-lfnm6.yaml) shows:

```
status:
  conditions:
  - lastTransitionTime: "2024-08-31T03:46:07Z"
    message: 0 of 2 completed
    reason: ContainerProvisioningFailed
    severity: Warning
    status: "False"
    type: Ready
  - lastTransitionTime: "2024-08-31T03:46:07Z"
    message: Re-provisioning
    reason: ContainerProvisioningFailed
    severity: Warning
    status: "False"
    type: ContainerProvisioned
```

---

## Example failure 2 (expected 2, found 1)

(`When testing clusterctl upgrades (v0.3=>v1.5=>current) Should create a management cluster and then upgrade all the providers `) [prow link](https://prow.k8s.io/view/gs/kubernetes-jenkins/logs/periodic-cluster-api-e2e-mink8s-main/1829718349949440000)

[capd controller logs](https://storage.googleapis.com/kubernetes-jenkins/logs/periodic-cluster-api-e2e-mink8s-main/1829718349949440000/artifacts/clusters/clusterctl-upgrade-management-7q5v4g/logs/capd-system/capd-controller-manager/capd-controller-manager-fb4b578f9-gq76c/manager.log) show: 

`I0831 04:16:20.583537       1 dockermachine_controller.go:220] controllers/DockerMachine/DockerMachine-controller "msg"="failed to create worker DockerMachine: timed out waiting for the condition, cleaning up so we can re-provision from a clean state" "cluster"="clusterctl-upgrade-workload-merjuc" "docker-cluster"="clusterctl-upgrade-workload-merjuc" "docker-machine"={"Namespace":"clusterctl-upgrade","Name":"clusterctl-upgrade-workload-merjuc-md-0-sr89b"} "machine"="clusterctl-upgrade-workload-merjuc-md-0-856f7cbb7c-fjfdl" `

in this failed test, we also see several failed creations of the control plane machine: 
`I0831 04:12:46.050065       1 dockermachine_controller.go:220] controllers/DockerMachine/DockerMachine-controller "msg"="failed to create worker DockerMachine: timed out waiting for the condition, cleaning up so we can re-provision from a clean state" "cluster"="clusterctl-upgrade-workload-merjuc" "docker-cluster"="clusterctl-upgrade-workload-merjuc" "docker-machine"={"Namespace":"clusterctl-upgrade","Name":"clusterctl-upgrade-workload-merjuc-control-plane-76jpw"} "machine"="clusterctl-upgrade-workload-merjuc-control-plane-kw497"`

most have the same error, the other CP machine creation failure looks like this: 
`I0831 04:13:25.895170       1 dockermachine_controller.go:220] controllers/DockerMachine/DockerMachine-controller "msg"="failed to create worker DockerMachine: command \"docker run --detach --tty --privileged --security-opt seccomp=unconfined --tmpfs /tmp --tmpfs /run --volume /var --volume /lib/modules:/lib/modules:ro --hostname clusterctl-upgrade-workload-merjuc-control-plane-kw497 --network kind --name clusterctl-upgrade-workload-merjuc-control-plane-kw497 --label io.x-k8s.kind.cluster=clusterctl-upgrade-workload-merjuc --label io.x-k8s.kind.role=control-plane --expose 44139 --volume=/var/run/docker.sock:/var/run/docker.sock --publish=127.0.0.1:44139:6443/TCP kindest/node:v1.22.17@sha256:9af784f45a584f6b28bce2af84c494d947a05bd709151466489008f80a9ce9d5\" failed with error: exit status 125, cleaning up so we can re-provision from a clean state" "cluster"="clusterctl-upgrade-workload-merjuc" "docker-cluster"="clusterctl-upgrade-workload-merjuc" "docker-machine"={"Namespace":"clusterctl-upgrade","Name":"clusterctl-upgrade-workload-merjuc-control-plane-76jpw"} "machine"="clusterctl-upgrade-workload-merjuc-control-plane-kw497"`

these appear to resolve and the provider ID gets set on the CP node: 
`I0831 04:15:55.173110       1 machine.go:337] controllers/DockerMachine/DockerMachine-controller "msg"="Setting Kubernetes node providerID" "cluster"="clusterctl-upgrade-workload-merjuc" "docker-cluster"="clusterctl-upgrade-workload-merjuc" "docker-machine"={"Namespace":"clusterctl-upgrade","Name":"clusterctl-upgrade-workload-merjuc-control-plane-76jpw"} "machine"="clusterctl-upgrade-workload-merjuc-control-plane-kw497"`

in this case, the MD DockerMachine is still [Waiting for BootstrapData](https://storage.googleapis.com/kubernetes-jenkins/logs/periodic-cluster-api-e2e-mink8s-main/1829718349949440000/artifacts/clusters/clusterctl-upgrade-management-7q5v4g/resources/clusterctl-upgrade/DockerMachine/clusterctl-upgrade-workload-merjuc-md-0-sr89b.yaml) 

the [CP DockerMachine](https://storage.googleapis.com/kubernetes-jenkins/logs/periodic-cluster-api-e2e-mink8s-main/1829718349949440000/artifacts/clusters/clusterctl-upgrade-management-7q5v4g/resources/clusterctl-upgrade/DockerMachine/clusterctl-upgrade-workload-merjuc-control-plane-76jpw.yaml) only transitioned to `Ready` only 12 seconds before the test timed out:
```
  conditions:
  - lastTransitionTime: "2024-08-31T04:15:56Z"
    status: "True"
    type: Ready
  - lastTransitionTime: "2024-08-31T04:15:56Z"
    status: "True"
    type: BootstrapExecSucceeded
```

## Example failure 3 (expected 2, found 0) 
`[It] When testing clusterctl upgrades (v0.3=>v1.5=>current) Should create a management cluster and then upgrade all the providers `: [prow link](https://prow.k8s.io/view/gs/kubernetes-jenkins/logs/periodic-cluster-api-e2e-main/1829446304858640384)

in this case, it seems that MD is waiting for the control plane (https://storage.googleapis.com/kubernetes-jenkins/logs/periodic-cluster-api-e2e-main/1829446304858640384/artifacts/clusters/clusterctl-upgrade-management-hywi70/resources/clusterctl-upgrade/DockerMachine/clusterctl-upgrade-workload-66mpos-md-0-ltqf6.yaml):

    status:
      conditions:
      - lastTransitionTime: "2024-08-30T10:01:44Z"
        message: 0 of 2 completed
        reason: WaitingForControlPlaneAvailable
        severity: Info
        status: "False"
        type: Ready
      - lastTransitionTime: "2024-08-30T10:01:44Z"
        reason: WaitingForControlPlaneAvailable
        severity: Info
        status: "False"
        type: ContainerProvisioned

looks like all the `Failed to create worker DockerMachine` errors in the CAPD controller logs in this case relate to the control plane: https://storage.googleapis.com/kubernetes-jenkins/logs/periodic-cluster-api-e2e-main/1829446304858640384/artifacts/clusters/clusterctl-upgrade-management-hywi70/logs/capd-system/capd-controller-manager/capd-controller-manager-fb4b578f9-rzpq9/manager.log:
  ```
  I0830 10:04:39.240063       1 dockermachine_controller.go:220] controllers/DockerMachine/DockerMachine-controller "msg"="failed to create worker DockerMachine: timed out waiting for the condition, cleaning up so we can re-provision from a clean state" "cluster"="clusterctl-upgrade-workload-66mpos" "docker-cluster"="clusterctl-upgrade-workload-66mpos" "docker-machine"={"Namespace":"clusterctl-upgrade","Name":"clusterctl-upgrade-workload-66mpos-control-plane-jbbxg"} "machine"="clusterctl-upgrade-workload-66mpos-control-plane-8kvwp" 
  ```
  - it does seem that eventually the CP machine comes up, but it's still bootstrapping when the test times out at `10:06:52.444`:
```
status:
  conditions:
  - lastTransitionTime: "2024-08-30T10:06:52Z"
    message: 1 of 2 completed
    reason: Bootstrapping
    severity: Info
    status: "False"
    type: Ready
  - lastTransitionTime: "2024-08-30T10:06:52Z"
    reason: Bootstrapping
    severity: Info
    status: "False"
    type: BootstrapExecSucceeded
  - lastTransitionTime: "2024-08-30T10:06:52Z"
    status: "True"
    type: ContainerProvisioned
```

### Label(s) to be applied

/kind flake
One or more /area label. See https://github.com/kubernetes-sigs/cluster-api/labels?q=area for the list of labels.

**Comment by cahillsf** [comment]
> Am I understanding this correctly that "cleaning up so we can re-provision from a clean state" is coming from Docker?

Yes from what i can tell

> To be honest, I'm not sure if we should just increase the timeout and then accept that we can only create Machines after multiple tries. This seems like just trying ignoring the root cause.

> I would prefer if we could try to correlate this https://github.com/kubernetes-sigs/cluster-api/issues/11133#issuecomment-2340263063 with the time when the number of flakes increased. If there is a correlation I would like to try to revert the concurrency change first.

yeah sure that sounds reasonable to me, will review again and try to pinpoint when this started happening (closed the timeout increase PR)

**Comment by sbueringer** [comment]
If we don't have clear data I would be definitely also fine with rolling back the concurrency change and then checking if that significantly improves the flake situation.

**Comment by cahillsf** [comment]
> If we don't have clear data I would be definitely also fine with rolling back the concurrency change and then checking if that significantly improves the flake situation.

This particular message appears to predate the concurrency increase.  Seems like it was present but quite rare up until 2024-07-31 when it begins to occur more frequently: https://storage.googleapis.com/k8s-triage/index.html?date=2024-08-07&text=Timed%20out%20waiting%20for%20all%20Machines%20to%20exist&job=.*periodic-cluster-api-e2e.*&test=.*clusterctl%20upgrades

**Comment by sbueringer** [comment]
So it happened before, but way more often since we increased concurrency?

**Comment by cahillsf** [comment]
> So it happened before, but way more often since we increased concurrency?

the trend seems to support it, but maybe not directly?

going back to [2024-08-19](https://storage.googleapis.com/k8s-triage/index.html?date=2024-08-19&text=Timed%20out%20waiting%20for%20all%20Machines%20to%20exist&job=.*periodic-cluster-api-e2e.*&test=.*clusterctl%20upgrades) before the concurrency change:
`2 clusters of 8 failures out of 137618 builds from 8/4/2024, 8:00:04 PM to 8/19/2024, 3:39:37 PM.` (0.005% frequency)

move to [2024-09-12](https://storage.googleapis.com/k8s-triage/index.html?date=2024-09-12&text=Timed%20out%20waiting%20for%20all%20Machines%20to%20exist&job=.*periodic-cluster-api-e2e.*&test=.*clusterctl%20upgrades) (a week after the concurrency change was backported to release-1.8):
`2 clusters of 46 failures (7 in last day) out of 141326 builds from 8/28/2024, 8:00:25 PM to 9/12/2024, 4:09:32 PM.` (0.03% frequency)

a bit further (the whole lookback window now encapsulates tests after the concurrency change) [2024-09-20](https://storage.googleapis.com/k8s-triage/index.html?date=2024-09-20&text=Timed%20out%20waiting%20for%20all%20Machines%20to%20exist&job=.*periodic-cluster-api-e2e.*&test=.*clusterctl%20upgrades):
`2 clusters of 71 failures (8 in last day) out of 221929 builds from 9/5/2024, 8:00:29 PM to 9/20/2024, 3:59:38 PM.` (0.03% frequency)

now today [2024-09-24](https://storage.googleapis.com/k8s-triage/index.html?date=2024-09-24&text=Timed%20out%20waiting%20for%20all%20Machines%20to%20exist&job=.*periodic-cluster-api-e2e.*&test=.*clusterctl%20upgrades) we have:
`2 clusters of 116 failures (13 in last day) out of 192878 builds from 9/9/2024, 8:00:27 PM to 9/24/2024, 8:01:45 AM.` (0.06% frequency)

**Comment by cahillsf** [comment]
> What do you think about reverting concurrency on main & release-1.8 for a week and then check if that gives us clear data?

sounds good to me

**Comment by sbueringer** [comment]
You'll open the PRs, I approve? :)

**Comment by sbueringer** [comment]
@cahillsf So far results seem to be pretty clear

![image](https://github.com/user-attachments/assets/fe4d4dfc-399a-4856-8605-d31d60e394f2)
> https://storage.googleapis.com/k8s-triage/index.html?job=.*periodic-cluster-api-e2e

**Comment by cahillsf** [comment]
yes they do indeed!  is there some other follow up work we can pursue to reintroduce the greater concurrency without these negative impacts?  or are we kind of "stuck" with 3 nodes?  

i guess before introducing more variables, it might be worth it to "refresh" our understanding of the flakes that exist now that we have a cleaner signal to see if some of them might be actual bugs versus side effects of resource constraints in the test infra.  what do you think is the best next step?

**Comment by sbueringer** [comment]
I think I would to keep the concurrency as is for now. Let's see what flakes remain and then try to fix them.

I'm also not sure if we can do much to increase the concurrency (maybe we can slightly increase it without getting new flakes but not much). It seems like we're simply hitting limits of the infrastructure when we run too much in parallel (IIRC we're not even close to hitting CPU limits though).

In general, we really need pretty stable CI the next 1-2 months until code freeze. We have a few critical things upcoming that could lead to some flakes that might not be easy to spot (so we need good visibility for new flakes :)).

**Comment by cahillsf** [comment]
@sbueringer taking another look at this message specifically, seems like we have 0 occurrences of this message since the concurrency was dropped back down: https://storage.googleapis.com/k8s-triage/index.html?date=2024-09-30&text=Timed%20out%20waiting%20for%20all%20Machines%20to%20exist&job=.*periodic-cluster-api-e2e.*&test=.*clusterctl%20upgrades

![Screenshot 2024-09-30 at 6 59 29 PM](https://github.com/user-attachments/assets/39c89165-80c5-4f99-b547-11df131932fb)

not sure if you want to give it some more time or we can just close this issue?

**Comment by sbueringer** [comment]
Fine for me to close

**Comment by cahillsf** [comment]
> Fine for me to close

cool -- thanks for all your help on this 🙇 

/close

**Comment by cahillsf** [comment]
hm, does appear to be some issue with the docker controller.  these flakes are quite rare (this pattern has again disappeared from recent testgrid view, but taking a look at this recent failure: https://prow.k8s.io/view/gs/kubernetes-ci-logs/logs/periodic-cluster-api-e2e-mink8s-release-1-7/1874453010868867072

we can see these big errors messages being captured in the capd controller logs: https://storage.googleapis.com/kubernetes-ci-logs/logs/periodic-cluster-api-e2e-mink8s-release-1-7/1874453010868867072/artifacts/clusters/clusterctl-upgrade-management-ycnwqg/logs/capd-system/capd-controller-manager/capd-controller-manager-5865b8bf8-56gws/manager.log

seems like the error is being thrown from the `kubeadm init` command:
```
\n\n\tUnfortunately, an error has occurred:\n\t\ttimed out waiting for the condition\n\n\tThis error is likely caused by:\n\t\t- The kubelet is not running\n\t\t- The kubelet is unhealthy due to a misconfiguration of the node in some way (required cgroups disabled)\n\n\tIf you are on a systemd-powered system, you can try to troubleshoot the error with the following commands:\n\t\t- 'systemctl status kubelet'\n\t\t- 'journalctl -xeu kubelet'\n\n\tAdditionally, a control plane component may have crashed or exited when started by the container runtime.
```

**Comment by chrischdi** [comment]
Let's check if this happens on the old provider (v0.3 or v0.4) or if this happens on the latest version (which is the one we want to test here)

/triage accepted

**Comment by chrischdi** [comment]
Same may apply here as in:
https://github.com/kubernetes-sigs/cluster-api/issues/11610#issuecomment-2607792687
